Library/Homebrew/unpack_strategy.rb

```diff
@@ -0,0 +1,340 @@
+  private_constant :MAX_MAGIC_NUMBER_LENGTH
+
+  def self.strategies
+    @strategies ||= [
+      JarUnpackStrategy,
+      ZipUnpackStrategy,
+      XarUnpackStrategy,
+      XzUnpackStrategy,
+      TarUnpackStrategy,
+      Bzip2UnpackStrategy,
+      GzipUnpackStrategy,
+      LzipUnpackStrategy,
+      GitUnpackStrategy,
+      MercurialUnpackStrategy,
+      SubversionUnpackStrategy,
+      CvsUnpackStrategy,
+      FossilUnpackStrategy,
+      BazaarUnpackStrategy,
+      P7ZipUnpackStrategy,
+      RarUnpackStrategy,
+      LhaUnpackStrategy,
+    ].freeze
+  end
+  private_class_method :strategies
+
+  def self.detect(path)
+    magic_number = if path.directory?
+      ""
+    else
+      File.binread(path, MAX_MAGIC_NUMBER_LENGTH)
+    end
+
+    strategy = strategies.detect do |s|
+      s.can_extract?(path: path, magic_number: magic_number)
+    end
+
+    # This is so that bad files produce good error messages.
+    strategy ||= case path.extname
+    when ".tar.gz", ".tgz", ".tar.bz2", ".tbz", ".tar.xz"
+      TarUnpackStrategy
+    when ".zip"
+      ZipUnpackStrategy
+    else
+      UncompressedUnpackStrategy
+    end
+
+    strategy.new(path)
+  end
+
+  attr_reader :path
+
+  def initialize(path)
+    @path = Pathname(path).expand_path
+  end
+
+  def extract(to: nil, basename: nil)
+    unpack_dir = Pathname(to || Dir.pwd).expand_path
+    unpack_dir.mkpath
+    extract_to_dir(unpack_dir, basename: basename)
+  end
+
+  def pipe_to_tar(tool, unpack_dir)
+    Utils.popen_read(tool, "-dc", path) do |rd|
+      Utils.popen_write("tar", "-x", "-f", "-", "-C", unpack_dir) do |wr|
+        buf = ""
+        wr.write(buf) while rd.read(16384, buf)
+      end
+    end
+  end
+  private :pipe_to_tar
+
+  def extract_to_dir(*)
+    raise NotImplementedError
+  end
+  private :extract_to_dir
+end
+
+class DirectoryUnpackStrategy < UnpackStrategy
+  def self.can_extract?(path:, **)
+    path.directory?
+  end
+
+  def extract_to_dir(unpack_dir, **)
+    FileUtils.cp_r path.children, unpack_dir, preserve: true
+  end
+  private :extract_to_dir
+end
+
+class UncompressedUnpackStrategy < UnpackStrategy
+  def self.can_extract?(**)
+    false
+  end
+
+  def extract_to_dir(unpack_dir, basename:)
+    FileUtils.cp path, unpack_dir/basename, preserve: true
+  end
+  private :extract_to_dir
+end
+
+class JarUnpackStrategy < UncompressedUnpackStrategy
+  def self.can_extract?(path:, magic_number:)
+    return false unless ZipUnpackStrategy.can_extract?(path: path, magic_number: magic_number)
+
+    # Check further if the ZIP is a JAR/WAR.
+    Open3.popen3("unzip", "-l", path) do |stdin, stdout, stderr, wait_thr|
+      stdin.close_write
+      stderr.close_read
+
+      begin
+        return stdout.each_line.any? { |l| l.match?(%r{\s+META-INF/MANIFEST.MF$}) }
+      ensure
+        stdout.close_read
+        wait_thr.kill
+      end
+    end
+  end
+end
+
+class P7ZipUnpackStrategy < UnpackStrategy
+  def self.can_extract?(magic_number:, **)
```

Comment:
Would be nicer to avoid ** here I think (and everywhere below).

Reply:
How would you avoid these?

Reply:
Specify the arguments that would be passed.

Reply:
Then RuboCop will complain about unused arguments.

Reply:
@reitermarkus They can be prefixed with `_` and there may be other options depending on the types of arguments (I could be wrong on this, though). I think that's preferable to accepting literally any number of random arguments.

Reply:
Ok, there is a RuboCop setting to allow unused keyword arguments.